src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,170 @@
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIsElementIndex(index);
+        return getNodeAtIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIsElementIndex(index);
+        Node<T> current = getNodeAtIndex(index);
+        T oldValue = current.item;
+        current.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIsElementIndex(index);
+        return unlink(getNodeAtIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        for (Node<T> current = first; current != null; current = current.next) {
+            if (current.item == object || current.item != null && current.item.equals(object)) {
+                unlink(current);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private T unlink(Node<T> node) {
+        final T element = node.item;
+        Node<T> next = node.next;
+        Node<T> prev = node.prev;
+
+        if (prev == null) {
+            first = next;
+        } else {
+            prev.next = next;
+            node.prev = null;
+        }
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+            node.next = null;
+        }
+        node.item = null;
+        size--;
+        return element;
+    }
+
+    private void linkMiddle(T element, Node<T> currentIndexNode) {
+        Node<T> prevNode = currentIndexNode.prev;
+        Node<T> newNode = new Node<>(prevNode, element, currentIndexNode);
+        currentIndexNode.prev = newNode;
+        if (prevNode == null) {
+            first = newNode;
+        } else {
+            prevNode.next = newNode;
+        }
+        size++;
+    }
+
+    private void linkLast(T element) {
+        Node<T> prevNode = last;
+        Node<T> newNode = new Node<>(prevNode, element, null);
+        last = newNode;
+        if (prevNode == null) {
+            first = newNode;
+        } else {
+            prevNode.next = newNode;
+        }
+        size++;
+    }
+
+    private void checkIndex(int index) {
+        if (!(index >= 0 && index <= size)) {
+            throw new IndexOutOfBoundsException("Index: "
+                    + index + " is out of bounds of size: " + size);
+        }
+    }
+
+    private void checkIsElementIndex(int index) {
+        if (!(index >= 0 && index < size)) {
+            throw new IndexOutOfBoundsException("Index: "
+                    + index + " is non existing index");
+        }
+    }
+
+    private Node<T> getNodeAtIndex(int index) {
+        Node<T> currentNode = first;
+        if (index < (size >> 1)) {
```

Comment:
this one is better for readability
```suggestion
        if (index < (size / 2)) {
```